File: ios/Talise/Features/Chat/AgentReceiptSheet.swift
import SwiftUI

/// A clean paper-style receipt for a successful agent transaction — a white
/// slip with a torn bottom edge, monospace ledger rows, a barcode, and a
/// "thank you". The same slip is rendered to an image so "Share receipt" hands
/// the system share sheet a crisp graphic.
struct AgentReceiptSheet: View {
    let amountUsd: Double
    let recipient: String
    let digest: String
    var title: String = "Sent"

    @Environment(\.dismiss) private var dismiss
    @State private var rendered: Image?

    private let ink = Color(red: 0.10, green: 0.13, blue: 0.09)   // near-black green
    private let paper = Color(red: 0.98, green: 0.98, blue: 0.96) // warm white
    private var suiscanURL: URL? { URL(string: "https://suiscan.xyz/mainnet/tx/\(digest)") }

    var body: some View {
        VStack(spacing: 20) {
            ScrollView(showsIndicators: false) {
                slip
                    .padding(.horizontal, 36)
                    .padding(.top, 24)
            }

            VStack(spacing: 12) {
                if let img = rendered {
                    ShareLink(item: img, preview: SharePreview("Talise receipt", image: img)) { shareLabel }
                        .buttonStyle(.plain)
                } else {
                    shareLabel.opacity(0.55)
                }
                if let url = suiscanURL {
                    Link(destination: url) {
                        Text("View on Suiscan")
                            .font(TaliseFont.body(13, weight: .medium))
                            .foregroundStyle(TaliseColor.fgMuted)
                    }
                }
            }
            .padding(.horizontal, 22)
            .padding(.bottom, 6)
        }
        .frame(maxWidth: .infinity)
        .background(TaliseColor.bg)
        .task { renderImage() }
    }

    private var shareLabel: some View {
        HStack(spacing: 8) {
            Image(systemName: "square.and.arrow.up").font(.system(size: 15, weight: .semibold))
            Text("Share receipt").font(TaliseFont.body(15, weight: .semibold))
        }
        .foregroundStyle(TaliseColor.bg)
        .frame(maxWidth: .infinity).frame(height: 50)
        .background(Capsule().fill(TaliseColor.greenMint))
    }

    /// The paper slip (also what we snapshot to share).
    private var slip: some View {
        VStack(spacing: 0) {
            VStack(spacing: 14) {
                // Brand mark (tinted to the receipt ink) instead of a text wordmark.
                Image("TaliseLogo")
                    .resizable()
                    .renderingMode(.template)
                    .aspectRatio(contentMode: .fit)
                    .frame(width: 34, height: 31)
                    .foregroundStyle(ink)
                    .padding(.top, 28)
                Text("PAYMENT RECEIPT")
                    .font(.system(size: 10, weight: .medium, design: .monospaced))
                    .tracking(3)
                    .foregroundStyle(ink.opacity(0.5))

                dashed

                VStack(spacing: 10) {
                    line(title.uppercased(), TaliseFormat.usd2(amountUsd))
                    line("TO", recipient)
                    line("DATE", shortDate)
                    line("NETWORK FEE", "FREE")
                    line("STATUS", "CONFIRMED")
                }
                .padding(.horizontal, 22)

                dashed

                HStack(alignment: .firstTextBaseline) {
                    Text("TOTAL").font(.system(size: 13, weight: .bold, design: .monospaced)).foregroundStyle(ink)
                    Spacer()
                    Text(TaliseFormat.usd2(amountUsd)).font(.system(size: 17, weight: .heavy, design: .monospaced)).foregroundStyle(ink)
                }
                .padding(.horizontal, 22)

                Barcode(seed: digest, ink: ink)
                    .frame(height: 46)
                    .padding(.horizontal, 22)
                    .padding(.top, 4)
                Text(shortDigest)
                    .font(.system(size: 10, weight: .regular, design: .monospaced))
                    .foregroundStyle(ink.opacity(0.55))

                Text("THANK YOU")
                    .font(.system(size: 12, weight: .semibold, design: .monospaced))
                    .tracking(3)
                    .foregroundStyle(ink.opacity(0.75))
                    .padding(.top, 6)
                    .padding(.bottom, 22)
            }
            .frame(maxWidth: .infinity)
            .background(paper)
            // Torn bottom edge.
            TornEdge().fill(paper).frame(height: 12)
        }
        .compositingGroup()
        .shadow(color: .black.opacity(0.35), radius: 18, x: 0, y: 10)
    }

    private func line(_ label: String, _ value: String) -> some View {
        HStack(alignment: .firstTextBaseline) {
            Text(label).font(.system(size: 11, weight: .regular, design: .monospaced)).foregroundStyle(ink.opacity(0.5))
            Spacer(minLength: 12)
            Text(value)
                .font(.system(size: 12, weight: .medium, design: .monospaced))
                .foregroundStyle(ink)
                .lineLimit(1).truncationMode(.middle)
        }
    }

    private var dashed: some View {
        Line().stroke(style: StrokeStyle(lineWidth: 1, dash: [3, 3]))
            .foregroundStyle(ink.opacity(0.25))
            .frame(height: 1)
            .padding(.horizontal, 22)
    }

    private var shortDigest: String {
        digest.count > 16 ? "\(digest.prefix(8))…\(digest.suffix(8))" : digest
    }
    private var shortDate: String {
        let f = DateFormatter(); f.dateFormat = "MMM d, yyyy  HH:mm"; return f.string(from: Date())
    }

    @MainActor private func renderImage() {
        let renderer = ImageRenderer(content: slip.frame(width: 320).padding(28).background(TaliseColor.bg))
        renderer.scale = 3
        if let ui = renderer.uiImage { rendered = Image(uiImage: ui) }
    }
}

// MARK: - Shapes

private struct Line: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path(); p.move(to: CGPoint(x: rect.minX, y: rect.midY)); p.addLine(to: CGPoint(x: rect.maxX, y: rect.midY)); return p
    }
}

/// A zigzag "torn paper" bottom edge.
private struct TornEdge: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        let teeth = 22
        let w = rect.width / CGFloat(teeth)
        p.move(to: CGPoint(x: rect.minX, y: rect.minY))
        p.addLine(to: CGPoint(x: rect.maxX, y: rect.minY))
        var x = rect.maxX
        var down = true
        while x > rect.minX {
            x -= w
            p.addLine(to: CGPoint(x: max(x, rect.minX), y: down ? rect.maxY : rect.minY))
            down.toggle()
        }
        p.closeSubpath()
        return p
    }
}

/// A faux barcode whose bar widths are derived from the tx digest (stable).
private struct Barcode: View {
    let seed: String
    let ink: Color
    var body: some View {
        Canvas { ctx, size in
            let bytes = Array(seed.utf8)
            guard !bytes.isEmpty else { return }
            var x: CGFloat = 0
            var i = 0
            while x < size.width {
                let b = bytes[i % bytes.count]
                let bar = CGFloat(1 + Int(b) % 4)       // 1...4 pt wide
                let gap = CGFloat(1 + Int(b >> 2) % 3)  // 1...3 pt gap
                if i % 2 == 0 {
                    ctx.fill(Path(CGRect(x: x, y: 0, width: bar, height: size.height)), with: .color(ink))
                }
                x += bar + gap
                i += 1
            }
        }
    }
}

File: android/app/src/main/java/io/talise/app/feature/chat/AgentReceiptSheet.kt
package io.talise.app.feature.chat

import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Share
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.talise.app.R
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * A clean paper-style receipt for a successful agent transaction — a white slip
 * with a torn bottom edge, monospace ledger rows, a barcode, and a "thank you".
 * The Android port of iOS `AgentReceiptSheet` (rendered inside a bottom sheet).
 * Sharing hands the system share sheet the receipt as text + the Suiscan link.
 */
@Composable
fun AgentReceiptSheet(
    amountUsd: Double,
    recipient: String,
    digest: String,
    title: String = "Sent",
) {
    val context = LocalContext.current
    val ink = Color(0xFF192117)   // near-black green
    val paper = Color(0xFFFAFAF5) // warm white
    val suiscanUrl = "https://suiscan.xyz/mainnet/tx/$digest"

    Column(
        Modifier
            .fillMaxWidth()
            .background(TaliseColors.bg)
            .padding(bottom = 24.dp),
        verticalArrangement = Arrangement.spacedBy(20.dp),
    ) {
        Column(
            Modifier
                .weight(1f, fill = false)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 36.dp)
                .padding(top = 24.dp),
        ) {
            Slip(amountUsd = amountUsd, recipient = recipient, digest = digest, title = title, ink = ink, paper = paper)
        }

        Column(
            Modifier.fillMaxWidth().padding(horizontal = 22.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            // Share receipt — hands the system share sheet the receipt summary.
            Row(
                Modifier
                    .fillMaxWidth()
                    .height(50.dp)
                    .clip(RoundedCornerShape(25.dp))
                    .background(TaliseColors.greenMint)
                    .clickable {
                        val share = Intent(Intent.ACTION_SEND).apply {
                            type = "text/plain"
                            putExtra(
                                Intent.EXTRA_TEXT,
                                "Talise receipt. $title ${AgentExecutor.usd2(amountUsd)} to $recipient. View on Suiscan: $suiscanUrl",
                            )
                        }
                        context.startActivity(Intent.createChooser(share, "Share receipt"))
                    },
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(Icons.Filled.Share, contentDescription = null, tint = TaliseColors.bg, modifier = Modifier.size(16.dp))
                Spacer(Modifier.width(8.dp))
                Text("Share receipt", style = TaliseType.body(15.sp, FontWeight.SemiBold), color = TaliseColors.bg)
            }
            Text(
                "View on Suiscan",
                style = TaliseType.body(13.sp, FontWeight.Medium),
                color = TaliseColors.fgMuted,
                modifier = Modifier.clickable {
                    context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(suiscanUrl)))
                },
            )
        }
    }
}

/** The paper slip itself. */
@Composable
private fun Slip(
    amountUsd: Double,
    recipient: String,
    digest: String,
    title: String,
    ink: Color,
    paper: Color,
) {
    Column(Modifier.fillMaxWidth().shadow(elevation = 18.dp, shape = RoundedCornerShape(2.dp))) {
        Column(
            Modifier.fillMaxWidth().background(paper),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(14.dp),
        ) {
            // Brand mark (tinted to the receipt ink) instead of a text wordmark.
            Image(
                painter = painterResource(R.drawable.taliselogo),
                contentDescription = null,
                colorFilter = ColorFilter.tint(ink),
                modifier = Modifier.padding(top = 28.dp).size(width = 34.dp, height = 31.dp),
            )
            Text(
                "PAYMENT RECEIPT",
                style = TaliseType.mono(10.sp, FontWeight.Medium).copy(letterSpacing = 3.sp),
                color = ink.copy(alpha = 0.5f),
            )

            DashedLine(ink)

            Column(
                Modifier.fillMaxWidth().padding(horizontal = 22.dp),
                verticalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                LedgerLine(title.uppercase(Locale.US), AgentExecutor.usd2(amountUsd), ink)
                LedgerLine("TO", recipient, ink)
                LedgerLine("DATE", shortDate(), ink)
                LedgerLine("NETWORK FEE", "FREE", ink)
                LedgerLine("STATUS", "CONFIRMED", ink)
            }

            DashedLine(ink)

            Row(
                Modifier.fillMaxWidth().padding(horizontal = 22.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text("TOTAL", style = TaliseType.mono(13.sp, FontWeight.Bold), color = ink)
                Spacer(Modifier.weight(1f))
                Text(AgentExecutor.usd2(amountUsd), style = TaliseType.mono(17.sp, FontWeight.Bold), color = ink)
            }

            Barcode(
                seed = digest,
                ink = ink,
                modifier = Modifier
                    .fillMaxWidth()
                    .height(46.dp)
                    .padding(horizontal = 22.dp)
                    .padding(top = 4.dp),
            )
            Text(shortDigest(digest), style = TaliseType.mono(10.sp), color = ink.copy(alpha = 0.55f))

            Text(
                "THANK YOU",
                style = TaliseType.mono(12.sp, FontWeight.SemiBold).copy(letterSpacing = 3.sp),
                color = ink.copy(alpha = 0.75f),
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(top = 6.dp, bottom = 22.dp),
            )
        }
        // Torn bottom edge.
        Spacer(
            Modifier
                .fillMaxWidth()
                .height(12.dp)
                .clip(tornEdgeShape())
                .background(paper),
        )
    }
}

@Composable
private fun LedgerLine(label: String, value: String, ink: Color) {
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        Text(label, style = TaliseType.mono(11.sp), color = ink.copy(alpha = 0.5f))
        Spacer(Modifier.width(12.dp).weight(1f))
        Text(
            value,
            style = TaliseType.mono(12.sp, FontWeight.Medium),
            color = ink,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

@Composable
private fun DashedLine(ink: Color) {
    Canvas(
        Modifier
            .fillMaxWidth()
            .height(1.dp)
            .padding(horizontal = 22.dp),
    ) {
        drawLine(
            color = ink.copy(alpha = 0.25f),
            start = Offset(0f, size.height / 2f),
            end = Offset(size.width, size.height / 2f),
            strokeWidth = 1.dp.toPx(),
            pathEffect = PathEffect.dashPathEffect(floatArrayOf(3.dp.toPx(), 3.dp.toPx())),
        )
    }
}

/** A faux barcode whose bar widths are derived from the tx digest (stable). */
@Composable
private fun Barcode(seed: String, ink: Color, modifier: Modifier = Modifier) {
    Canvas(modifier) {
        val bytes = seed.toByteArray(Charsets.UTF_8)
        if (bytes.isEmpty()) return@Canvas
        var x = 0f
        var i = 0
        val unit = 1.dp.toPx()
        while (x < size.width) {
            val b = bytes[i % bytes.size].toInt() and 0xFF
            val bar = (1 + b % 4) * unit
            val gap = (1 + (b shr 2) % 3) * unit
            if (i % 2 == 0) {
                drawRect(
                    color = ink,
                    topLeft = Offset(x, 0f),
                    size = androidx.compose.ui.geometry.Size(bar, size.height),
                )
            }
            x += bar + gap
            i += 1
        }
    }
}

/** A zigzag "torn paper" bottom edge. */
private fun tornEdgeShape(): Shape = androidx.compose.foundation.shape.GenericShape { size, _ ->
    val teeth = 22
    val w = size.width / teeth
    moveTo(0f, 0f)
    lineTo(size.width, 0f)
    var x = size.width
    var down = true
    while (x > 0f) {
        x -= w
        lineTo(maxOf(x, 0f), if (down) size.height else 0f)
        down = !down
    }
    close()
}

private fun shortDigest(digest: String): String =
    if (digest.length > 16) "${digest.take(8)}…${digest.takeLast(8)}" else digest

private fun shortDate(): String =
    SimpleDateFormat("MMM d, yyyy  HH:mm", Locale.US).format(Date())
